host/xtest/pkcs11_1000.c

```diff
@@ -1463,3 +1483,492 @@ static void xtest_pkcs11_test_1007(ADBG_Case_t *c)
+		destroy_object = 1;
+
+		rv = C_CreateObject(session, test->attr_key, test->attr_count,
+				    &key_handle);
+
+		if (!ADBG_EXPECT_CK_OK(c, rv))
+			goto out;
+
+		/* Test signature in 1 step */
+		if (test->in != NULL) {
+			rv = C_SignInit(session, test->mechanism, key_handle);
+
+			if (!ADBG_EXPECT_CK_OK(c, rv))
+				goto out;
+
+			rv = C_SignUpdate(session,
+					  (void *)test->in, test->in_len);
+
+			if (!ADBG_EXPECT_CK_OK(c, rv))
+				goto out;
+
+			/* Test too short buffer case */
+			out_size = 1;
+			rv = C_SignFinal(session, out, &out_size);
+
+			if (!ADBG_EXPECT_CK_RESULT(c, CKR_BUFFER_TOO_SMALL, rv))
+				goto out;
+
+			/*
+			 * Test NULL buffer case with size as 0
+			 * to get the out_size
+			 */
+			out_size = 0;
+			rv = C_SignFinal(session, NULL, &out_size);
+
+			if (!ADBG_EXPECT_CK_OK(c, rv))
+				goto out;
+
+			/* Get to full output */
+			memset(out, 0, out_size);
+			rv = C_SignFinal(session, out, &out_size);
+
+			if (!ADBG_EXPECT_CK_OK(c, rv))
+				goto out;
+
+			(void)ADBG_EXPECT_BUFFER(c, test->out,
+						 test->out_len,
+						 out, out_size);
+		}
+
+		/* Test 2 step update signature */
+		rv = C_SignInit(session, test->mechanism, key_handle);
+
+		if (!ADBG_EXPECT_CK_OK(c, rv))
+			goto out;
+
+		if (test->in != NULL) {
+			rv = C_SignUpdate(session,
+					  (void *)test->in, test->in_incr);
+
+			if (!ADBG_EXPECT_CK_OK(c, rv))
+				goto out;
+
+			rv = C_SignUpdate(session,
+					 (void *)(test->in + test->in_incr),
+					 test->in_len - test->in_incr);
+
+			if (!ADBG_EXPECT_CK_OK(c, rv))
+				goto out;
+		}
+
+		out_size = sizeof(out);
+		memset(out, 0, sizeof(out));
+
+		rv = C_SignFinal(session, out, &out_size);
+
+		if (!ADBG_EXPECT_CK_OK(c, rv))
+			goto out;
+
+		(void)ADBG_EXPECT_BUFFER(c, test->out,
+					 test->out_len, out, out_size);
+
+		/* Test 3 signature in one shot */
+		if (test->in != NULL) {
+			rv = C_SignInit(session, test->mechanism, key_handle);
+
+			if (!ADBG_EXPECT_CK_OK(c, rv))
+				goto out;
+
+			/* Test too short buffer case */
+			out_size = 1;
+			rv = C_Sign(session,
+				    (void *)test->in, test->in_len,
+				    out, &out_size);
+
+			if (!ADBG_EXPECT_CK_RESULT(c, CKR_BUFFER_TOO_SMALL, rv))
+				goto out;
+
+			/*
+			 * Test NULL buffer case with size as 0
+			 * to get the out_size
+			 */
+			out_size = 0;
+			rv = C_Sign(session,
+				    (void *)test->in, test->in_len,
+				    NULL, &out_size);
+
+			if (!ADBG_EXPECT_CK_OK(c, rv))
+				goto out;
+
+
+			/* Get to full output */
+			memset(out, 0, out_size);
+			rv = C_Sign(session,
+				    (void *)test->in, test->in_len,
+				    out, &out_size);
+
+			if (!ADBG_EXPECT_CK_OK(c, rv))
+				goto out;
+
+			(void)ADBG_EXPECT_BUFFER(c, test->out,
+						 test->out_len,
+						 out, out_size);
+		}
+
+		destroy_object = 0;
+		rv = C_DestroyObject(session, key_handle);
+
+		if (!ADBG_EXPECT_CK_OK(c, rv))
+			goto out;
+
+		Do_ADBG_EndSubCase(c, NULL);
+		close_subcase = 0;
+	}
+out:
+	if (close_subcase)
```

Comment:
I find these extra state variables slightly messy. We're normally using the kernel style of gotos to the right spot in this cleanup part of the function. With that this section could be like:
``` C
        goto out;
err:
         Do_ADBG_EndSubCase(c, NULL);
err_destr_obj:
         ADBG_EXPECT_CK_OK(c, C_DestroyObject(session, key_handle));
out:
         ADBG_EXPECT_CK_OK(c, C_CloseSession(session));
         ADBG_EXPECT_CK_OK(c, close_lib());
}
```

Reply:
I will change it.